### PR TITLE
TYP: Make mypy 0.800 compatible

### LINKED ISSUE #39399: TYP upgrade and pin mypy to 0.800
mypy 0.800 came out 3 days ago https://mypy-lang.blogspot.com/2021/01/mypy-0800-released.html , so it would be good to try upgrading

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #39399
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

### CODE REVIEW

**Comment by jbrockmendel** [comment]
LGTM cc @simonjayhawkins

File: environment.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@ dependencies:
   - flake8
   - flake8-comprehensions>=3.1.0  # used by flake8, linting of unnecessary comprehensions
   - isort>=5.2.1  # check that imports are in the right order
-  - mypy=0.790
+  - mypy
```
Comment (jreback) [inline environment.yml:26]: i think ok to pin >=

File: pandas/plotting/_matplotlib/boxplot.py [hunk 2]
```diff
@@ -28,7 +28,7 @@ class BoxPlot(LinePlot):
 
     _valid_return_types = (None, "axes", "dict", "both")
     # namedtuple to hold results
-    BP = namedtuple("Boxplot", ["ax", "lines"])
+    BP = namedtuple("BP", ["ax", "lines"])
```
Comment (jreback) [inline pandas/plotting/_matplotlib/boxplot.py:31]: OT we could use data classes here

File: pandas/tests/io/parser/common/test_file_buffer_url.py [hunk 3]
```diff
@@ -330,8 +331,9 @@ def test_read_csv_file_handle(all_parsers, io_class, encoding):
     parser = all_parsers
     expected = DataFrame({"a": [1], "b": [2]})
 
-    content = "a,b\n1,2"
+    content: Union[str, bytes] = "a,b\n1,2"
```
Comment (simonjayhawkins) [inline pandas/tests/io/parser/common/test_file_buffer_url.py:334]:
```suggestion
    content: str | bytes = "a,b\n1,2"
```

:grin:

we have 
```
[mypy-pandas.tests.*]
check_untyped_defs=False
```
so not sure why mypy is checking this function anyway.

OTOH could refactor...

```
    content = "a,b\n1,2"
    handle = io_class(content.encode("utf-8") if io_class == BytesIO else content)
```

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @phofl generally lgtm

File: environment.yml (hunk 1 above)
Comment (phofl) [reply environment.yml:26]: Thx, pinned again

File: pandas/util/_decorators.py [hunk 4]
```diff
@@ -79,7 +79,7 @@ def wrapper(*args, **kwargs) -> Callable[..., Any]:
         {dedent(doc)}"""
         )
 
-    return wrapper
+    return wrapper  # type: ignore[return-value]
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/util/_decorators.py:82]: Done

File: pandas/io/stata.py [hunk 5]
```diff
@@ -2491,9 +2491,9 @@ def write_file(self) -> None:
                 self.handles.close()
                 # Only @runtime_checkable protocols can be used with instance and class
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/io/stata.py:2492]: Done

File: pandas/io/common.py [hunk 6]
```diff
@@ -182,7 +182,7 @@ def stringify_path(
         return cast(FileOrBuffer[AnyStr], filepath_or_buffer)
 
     # Only @runtime_checkable protocols can be used with instance and class checks
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/io/common.py:184]: Done

File: pandas/tests/io/parser/common/test_file_buffer_url.py (hunk 3 above)
Comment (phofl) [reply pandas/tests/io/parser/common/test_file_buffer_url.py:334]: Hm the refactor is better. Thx.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @phofl lgtm pending green

**Comment by phofl** [comment]
@simonjayhawkins green.

Could you point me to an example of something like ``str | bytes``?

**Comment by simonjayhawkins** [comment]
> Could you point me to an example of something like `str | bytes`? Searched for something like this but could not find it

from http://mypy-lang.blogspot.com/

Two new Python features improve this situation and are now supported by mypy:

PEP 585 lets you use list[int] instead of List[int] (no need to import List and other generic collections from typing).
PEP 604 lets you write X | Y instead of Union[X, Y], and X | None instead of Optional[X] (no need to import Union or Optional from typing).
Note: Using list[int] requires Python 3.9 and X | Y requires Python 3.10 (alpha) in order to work at runtime. To use them on older versions of Python, use from \_\_future__ import annotations. This allows them to be used in type annotations, but the older variants (or string literal escaping) may be required in non-annotation contexts, such as in type aliases. See the docs for more details.

**Comment by phofl** [comment]
thanks very much, should have looked there and not in our codebase :)

**Comment by simonjayhawkins** [comment]
@phofl also need to update the version in the doc/source/whatsnew/v1.3.0.rst

**Comment by phofl** [comment]
Thx, did not think about this. #39419

**Pull request merged**